### PR TITLE
*: Support record stmtsummary evicted count

### PR DESCRIPTION
**Pull request by Max-Cheng**
### What is changed and how it works?

What's Changed:Add a new table `statement_summary_evicted` to record `statement_summary` evicted data

How it Works:Monitor the length of the LRU to make sure last element will be evicted

### Related changes

- PR to update `executor/builder.go`/`executor/infoschema_reader.go`/`infoschema/perfschema/tables.go`/`infoschema/tables.go`:
- Need to cherry-pick to the release branch

### Check List 
- Manual test (add detailed scripts or steps below)

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM

### Release note 

- Support record statement_summary evicted data

### CODE REVIEW

File: util/stmtsummary/stmtevicted_summary.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+	Count	int64
+	historyMax int
+}
+
+func newstmtEvictedMap() *stmtEvictedByDigestMap {
+	return &stmtEvictedByDigestMap{
+		//Only record 96 time window(48H)
+		//maybe we can modify this later
+		summaryMap:              		kvcache.NewSimpleLRUCache(96,0,0),
+		beginTimeForCurInterval:		900,
```
Comment (crazycs520) [inline util/stmtsummary/stmtevicted_summary.go:44]: Don't use const.
The capacity should keep consistant with the config of `stmtSummaryByDigestMap `.

File: util/stmtsummary/stmtevicted_summary.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+	// beginTimeForCurInterval is the begin time for current summary.
+	beginTimeForCurInterval int64
+}
+
+type stmEvictedByDigestKey struct {
+	hash          []byte
+	beginTimeHash string
+}
+
+type stmEvictedByDigest struct {
```
Comment (crazycs520) [inline util/stmtsummary/stmtevicted_summary.go:29]: This is not group by `digest`, how about rename to `stmtSummaryEvictedInfo`

File: util/stmtsummary/stmtevicted_summary.go (hunk 1 above)
Comment (Max-Cheng) [reply util/stmtsummary/stmtevicted_summary.go:44]: 👌🏻

File: util/stmtsummary/stmtevicted_summary.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	eb.Lock()
+	defer eb.Unlock()
+	if !eb.isInit{
+		eb.init(begintime,begintime+900)
+	}
+	EvictedCol:=value.(*stmtSummaryByDigest)
+	for i:=EvictedCol.history.Front();i!=nil;i=i.Next(){
+		if i.Value.(*stmtSummaryByDigestElement).endTime>begintime {
+			eb.history.PushBack(i.Value.(*stmtSummaryByDigestElement))
+			//Add Count
```
Comment (jyz0309) [inline util/stmtsummary/stmtevicted_summary.go:101]: ditto

**Comment by Max-Cheng** [comment]
> /run-unit-test

I will push newest version at tonight then we can find some weird things……

**Comment by Max-Cheng** [comment]
@jyz0309 👀

File: executor/infoschema_reader.go [hunk 4]
```diff
@@ -1829,6 +1831,10 @@ func (e *memtableRetriever) setDataForStatementsSummary(ctx sessionctx.Context,
 		e.rows = stmtsummary.StmtSummaryByDigestMap.ToHistoryDatum(user, isSuper)
 	}
 	switch tableName {
+	case infoschema.TableStatementsSummaryEvicted:
+		e.rows = stmtsummary.StmtEvictedMap.ToCurrentDatum()
+	}
+	switch tableName {
```
Comment (jyz0309) [inline executor/infoschema_reader.go:1837]: Why we need 2 switch tableName here-.-

File: util/stmtsummary/stmtevicted_summary.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+	eb.endTime = endtime
+	eb.historyMax = historysize
+	eb.Count = 0
+}
+
+func (eb *stmtSummaryEvictedInfo) toDatum() []types.Datum {
+	eb.Lock()
+	defer eb.Unlock()
+	return types.MakeDatums(
+		types.NewTime(types.FromGoTime(time.Unix(eb.beginTime, 0)), mysql.TypeTimestamp, 0),
```
Comment (jyz0309) [inline util/stmtsummary/stmtevicted_summary.go:129]: maybe we can use atomic.load to load the `beginTime` and `endTime` instead  of the lock ? I think use lock here will make a large performance waste?

Comment (jyz0309) [reply util/stmtsummary/stmtevicted_summary.go:129]: It's only my personal suggestion. I think you need to think about that use sync.Mutex here is better than use atomic or not.

**Comment by jyz0309** [comment]
Of course, plz add some test~

**Comment by Max-Cheng** [comment]
@jyz0309 @crazycs520 I'm still stuck in `infoschema/tables_test.go`, and I still can't understand how the relevant code in the file works.

**Comment by Max-Cheng** [comment]
> @jyz0309 @crazycs520 I'm still stuck in `infoschema/tables_test.go`, and I still can't understand how the relevant code in the file works.

When I try to add system variables, the test for cluster information never passes🥲

**Comment by Max-Cheng** [comment]
The local test keeps failing, but the push passes😳

**Comment by tangenta** [comment]
@Max-Cheng Before you run `make dev`, you need to set the proxy to empty:
```bash
unproxy () {
	unset HTTP_PROXY
	unset HTTPS_PROXY
	unset http_proxy
	unset https_proxy
}
```

Where did you see the tutorial that guides local testing? Perhaps we can improve it:)

**Comment by Max-Cheng** [comment]
> @Max-Cheng Before you run `make dev`, you need to set the proxy to empty:
> 
> ```shell
> unproxy () {
> 	unset HTTP_PROXY
> 	unset HTTPS_PROXY
> 	unset http_proxy
> 	unset https_proxy
> }
> ```
> 
> Where did you see the tutorial that guides local testing? Perhaps we can improve it:)

Nah,`infoschema/tables_test.go` can't pass `L826&L770` But thanks for your suggest

**Pull request closed without merging**